tests/Pgvector.Tests/EntityFrameworkCoreTests.cs

```diff
@@ -34,28 +34,177 @@ public class Item
-        ctx.Items.Add(new Item { Embedding = new Vector(new float[] { 2, 2, 2 }) });
-        ctx.Items.Add(new Item { Embedding = new Vector(new float[] { 1, 1, 2 }) });
-        ctx.SaveChanges();
+        var itemsB = await db.Items
+            .OrderBy(x => x.Embedding!.L2Distance(embedding))
+            .Take(5)
+            .ToListAsync();
+
+        foreach (Item item in itemsB)
+        {
+            if (item.Embedding != null)
+            {
+                Console.WriteLine(item.Embedding);
+            }
+        }
+
+        Assert.Equal(itemsA.Count, itemsB.Count); // Check amount
+        Assert.Equal(itemsA.Select(x => x.Id).ToArray(), itemsB.Select(x => x.Id).ToArray()); // Check order
+    }
+
+    [Fact]
+    public async Task CosineDistanceSelectOrdered()
+    {
+        var db = _fixture.Db;
 
         var embedding = new Vector(new float[] { 1, 1, 1 });
-        var items = await ctx.Items.FromSql($"SELECT * FROM efcore_items ORDER BY embedding <-> {embedding} LIMIT 5").ToListAsync();
-        foreach (Item item in items)
+
+        var itemsA = await db.Items
+            .FromSql($"SELECT * FROM efcore_items ORDER BY embedding <=> {embedding} LIMIT 5")
+            .ToListAsync();
+
+        var itemsB = await db.Items
+            .OrderBy(x => x.Embedding!.CosineDistance(embedding))
+            .Take(5)
+            .ToListAsync();
+
+        foreach (Item item in itemsB)
         {
             if (item.Embedding != null)
             {
                 Console.WriteLine(item.Embedding);
             }
         }
+
+        Assert.Equal(itemsA.Count, itemsB.Count); // Check amount
+        Assert.Equal(itemsA.Select(x => x.Id).ToArray(), itemsB.Select(x => x.Id).ToArray()); // Check order
+    }
+
+    [Fact]
+    public async Task MaxInnerProductSelectOrdered()
+    {
+        var db = _fixture.Db;
+
+        var embedding = new Vector(new float[] { 1, 1, 1 });
+
+        var itemsA = await db.Items
+            .FromSql($"SELECT * FROM efcore_items ORDER BY embedding <#> {embedding} LIMIT 5")
+            .ToListAsync();
+
+        var itemsB = await db.Items
+            .OrderBy(x => x.Embedding!.MaxInnerProduct(embedding))
+            .Take(5)
+            .ToListAsync();
+
+        foreach (Item item in itemsB)
+        {
+            if (item.Embedding != null)
+            {
+                Console.WriteLine(item.Embedding);
+            }
+        }
+
+        Assert.Equal(itemsA.Count, itemsB.Count); // Check amount
+        Assert.Equal(itemsA.Select(x => x.Id).ToArray(), itemsB.Select(x => x.Id).ToArray()); // Check order
+    }
+
+    [Fact]
+    public void L2DistanceQuery()
+    {
+        var db = _fixture.Db;
+
+        var embedding = new Vector(new float[] { 1, 1, 1 });
+
+        var query = db.Items
+            .OrderBy(x => x.Embedding!.L2Distance(embedding))
+            .Take(5);
+
+        var queryString = query.ToQueryString();
```

Comment:
Any reason to do ToQueryString rather than execute the query (ToListAsync) and assert the results, as is being done above in the SQL tests? This needlessly reduces the coverage of the test, not checking that the query actually works etc.

Reply:
I assumed it would be useful to provide better coverage of the new features.

Is it really necessary or not?

Reply:
I think you misunderstood - I absolutely agree that there should be tests with queries that use the new functions you've introduced (I don't think they should be removed as you've done in the PR).

What I wrote was that these tests should actually *execute* the queries against the database, rather than calling ToQueryString and asserting the result SQL. Only doing ToQueryString means that if the SQL is wrong and wouldn't actually work (e.g. since you accidentally got the operator wrong), the test wouldn't uncover it.

Reply:
@gluschenko I see that the tests are back, but they're still doing ToQueryString.

Reply:
Sure, I've reverted my last commit. Today I'm going to do this thing more correct.

Am I correct in understanding that I need to run the query in this way? Or should I not use `ToQueryString()` at all and take raw SQL from the logs?

```
public async Task CosineDistanceQuery()
{
    var db = _fixture.Db;

    var embedding = new Vector(new float[] { 1, 1, 1 });

    var query = db.Items
        .OrderBy(x => x.Embedding!.CosineDistance(embedding))
        .Take(5);

    var items = await query.ToArrayAsync();  // <-----
    Assert.True(items.Any());                // <-----

    var queryString = query.ToQueryString();

    Console.WriteLine(queryString);

    Assert.Contains("ORDER BY e.embedding <=> @__embedding_0", queryString);
}
```

Reply:
Yes, something like that. You should assert that the items coming out of ToArrayAsync are the actual expected ones (as opposed to just any items come out; note how this is done in the raw SQL tests above - you should be doing something similar.

Regarding asserting on the SQL - doing it as you are doing with ToQueryString is OK. In the EF tests we hook up logging and use that to assert SQL, but that's likely too much effort for what's being tested here.

Reply:
I did it in the new commit ef5e17a. Each of query validation tests now has a `ToArrayAsync()` call.

Reply:
Your tests are still only asserting that **some** items are being returned:

```c#
var items = await query.ToArrayAsync();
Assert.True(items.Any());
```

It would be better to assert that the specific expected ones are returned instead, as the already existing SQL-tests above do.

Reply:
Is that enough? 😁

```
var items = await query.ToArrayAsync();
Assert.Equal(5, items.Length);
```

Reply:
Or do we need to specifically check the order and specific item ids?

```csharp
var items = await query.ToArrayAsync();
Assert.Equal(new[] { 11, 10, 9, 8, 7 }, items.Select(x => x.Id).ToArray());
```

Reply:
Ideally assert the exact result coming back from the database, as the existing tests already do. This provides full coverage.

Reply:
The query tests are now assert exact `Vector` values, rather than the `id` column. Also, I have overridden `Equals` and `GetHashCode` on `Vector` class.

```csharp
Assert.Equal(
    new[] 
    {
        new Vector(new float[] { 1, 1, 1 }),
        new Vector(new float[] { 2, 2, 2 }),
        new Vector(new float[] { 0, 0, 0 }),
        new Vector(new float[] { 3, 3, 3 }),
        new Vector(new float[] { -1, -1, -1 })
    }, 
    items.Select(x => x.Embedding).ToArray()
);
```